rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,275 @@
+| FlexBuffers       | 401    | 855        | 23256    | 264      | 490          | 309         |
+| JSON              | 550    | 438        | 544      | 396      | 439          | 184         |
+| BSON              | 115    | 103        | 109      | 80       | 414          | 216         |
+| MessagePack       | 135    | 222        | 237      | 119      | 296          | 187         |
+| **Compiled Libs** |        |            |          |          |              |             |
+| Flatbuffers       | 1046   | 14706      | 250000   | 1065     | 264          | 181         |
+| Bincode           | 5882   | 8772       | 9524     | 4016     | 163          | 129         |
+| Protobuf          | 859    | 1140       | 1163     | 480      | 154          | 141         |
+| Prost             | 1225   | 1866       | 1984     | 962      | 154          | 142         |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks). Some of these formats (JSON for example) still leaves us with a marginal concern around dropped metadata.
+
+While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. We have a few specific domains of concern around the chosen path:
+
+- Performance
+- Compatibility with Transports
+- Maintaining backcompat
+- Tooling and Maintenance costs
+
+These four concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to keep in mind. So, to reiterate - our ideal solution solves for all four but that might just not be possible.
+
+It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MessagePack or CBOR. And their maintenance cost has already been paid since we're using them today. Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.
+
+With regards to the tansport and the suggested path of implementing gRPC we have to keep both perf and kubernetes issues in mind. [TCP causes us some problems in our K8s integration today](https://github.com/timberio/vector/issues/2070) and unfortunately this problem has quite a bit more context to it than can sanely be shared in this RFC. Suffice to say that any choice we make here has repurcussions on our deployment architecture and ultimately gRPC with Protobufs provides us what we think are the right tradeoffs for our implementation. 
```

Comment:
Woof, yes. We're not the only people that have these issues but k8s load balancing makes for a strange environment to do gRPC in. _Because_ we're not the only ones with these issues I'd expect resolution eventually -- running headless or inside a service mesh seem to be common now -- but it's for sure a foot gun for users.

Reply:
OH YEAH. @MOZGIII and I had discussed this at length when I originally wrote the RFC and indeed there will be some problems to solve there no matter what. We both kind of felt like aligning behind GRPC and Protobuf should give us the most likely guarantee that aspects of this specific problem soften as time goes on for that exact reason!